File: lite/examples/model_personalization/android/app/src/main/java/org/tensorflow/lite/examples/transfer/ImageUtils.java
/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

package org.tensorflow.lite.examples.transfer;

/** Utility class for manipulating images.
 *
 * Copy-pasted from TFLite object detection example.
 */
public class ImageUtils {
  // TODO(b/137790961) reduce code duplication between different TFLite examples.

  // This value is 2 ^ 18 - 1, and is used to clamp the RGB values before their ranges
  // are normalized to eight bits.
  static final int MAX_CHANNEL_VALUE = 262143;

  public static void convertYUV420SPToARGB8888(byte[] input, int width, int height, int[] output) {
    final int frameSize = width * height;
    for (int j = 0, yp = 0; j < height; j++) {
      int uvp = frameSize + (j >> 1) * width;
      int u = 0;
      int v = 0;

      for (int i = 0; i < width; i++, yp++) {
        int y = 0xff & input[yp];
        if ((i & 1) == 0) {
          v = 0xff & input[uvp++];
          u = 0xff & input[uvp++];
        }

        output[yp] = yuv2Rgb(y, u, v);
      }
    }
  }

  private static int yuv2Rgb(int y, int u, int v) {
    // Adjust and check YUV values
    y = (y - 16) < 0 ? 0 : (y - 16);
    u -= 128;
    v -= 128;

    // This is the floating point equivalent. We do the conversion in integer
    // because some Android devices do not have floating point in hardware.
    // nR = (int)(1.164 * nY + 2.018 * nU);
    // nG = (int)(1.164 * nY - 0.813 * nV - 0.391 * nU);
    // nB = (int)(1.164 * nY + 1.596 * nV);
    int y1192 = 1192 * y;
    int r = (y1192 + 1634 * v);
    int g = (y1192 - 833 * v - 400 * u);
    int b = (y1192 + 2066 * u);

    // Clipping RGB values to be inside boundaries [ 0 , MAX_CHANNEL_VALUE ]
    r = r > MAX_CHANNEL_VALUE ? MAX_CHANNEL_VALUE : Math.max(r, 0);
    g = g > MAX_CHANNEL_VALUE ? MAX_CHANNEL_VALUE : Math.max(g, 0);
    b = b > MAX_CHANNEL_VALUE ? MAX_CHANNEL_VALUE : Math.max(b, 0);

    return 0xff000000 | ((r << 6) & 0xff0000) | ((g >> 2) & 0xff00) | ((b >> 10) & 0xff);
  }

  public static void convertYUV420ToARGB8888(
      byte[] yData,
      byte[] uData,
      byte[] vData,
      int width,
      int height,
      int yRowStride,
      int uvRowStride,
      int uvPixelStride,
      int[] out) {
    int yp = 0;
    for (int j = 0; j < height; j++) {
      int pY = yRowStride * j;
      int pUV = uvRowStride * (j >> 1);

      for (int i = 0; i < width; i++) {
        int uvOffset = pUV + (i >> 1) * uvPixelStride;

        out[yp++] = yuv2Rgb(0xff & yData[pY + i], 0xff & uData[uvOffset], 0xff & vData[uvOffset]);
      }
    }
  }
}

File: lite/examples/posenet/android/app/src/main/java/org/tensorflow/lite/examples/posenet/ImageUtils.kt
/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

package org.tensorflow.lite.examples.posenet

/** Utility class for manipulating images.  */
object ImageUtils {
  // This value is 2 ^ 18 - 1, and is used to hold the RGB values together before their ranges
  // are normalized to eight bits.
  private const val MAX_CHANNEL_VALUE = 262143

  /** Helper function to convert y,u,v integer values to RGB format */
  private fun convertYUVToRGB(y: Int, u: Int, v: Int): Int {
    // Adjust and check YUV values
    val yNew = if (y - 16 < 0) 0 else y - 16
    val uNew = u - 128
    val vNew = v - 128
    val expandY = 1192 * yNew
    var r = expandY + 1634 * vNew
    var g = expandY - 833 * vNew - 400 * uNew
    var b = expandY + 2066 * uNew

    // Clipping RGB values to be inside boundaries [ 0 , MAX_CHANNEL_VALUE ]
    val checkBoundaries = { x: Int ->
      when {
        x > MAX_CHANNEL_VALUE -> MAX_CHANNEL_VALUE
        x < 0 -> 0
        else -> x
      }
    }
    r = checkBoundaries(r)
    g = checkBoundaries(g)
    b = checkBoundaries(b)
    return -0x1000000 or (r shl 6 and 0xff0000) or (g shr 2 and 0xff00) or (b shr 10 and 0xff)
  }

  /** Converts YUV420 format image data (ByteArray) into ARGB8888 format with IntArray as output. */
  fun convertYUV420ToARGB8888(
    yData: ByteArray,
    uData: ByteArray,
    vData: ByteArray,
    width: Int,
    height: Int,
    yRowStride: Int,
    uvRowStride: Int,
    uvPixelStride: Int,
    out: IntArray
  ) {
    var outputIndex = 0
    for (j in 0 until height) {
      val positionY = yRowStride * j
      val positionUV = uvRowStride * (j shr 1)

      for (i in 0 until width) {
        val uvOffset = positionUV + (i shr 1) * uvPixelStride

        // "0xff and" is used to cut off bits from following value that are higher than
        // the low 8 bits
        out[outputIndex] = convertYUVToRGB(
          0xff and yData[positionY + i].toInt(), 0xff and uData[uvOffset].toInt(),
          0xff and vData[uvOffset].toInt()
        )
        outputIndex += 1
      }
    }
  }
}
